[Core] Refactor CommandLine arguments

ltpquang opened the pull request:
### Problem

I'm about to call Cucumber `main` function from my own app

```
public static void main(String[] args) {
        io.cucumber.core.cli.Main.main(args);
}
```

For better management, I would standardize my app's arguments, before passing to Cucumber.

The problem is Cucumber's arguments are now hardcoded, which make future updates maybe break my app.

### Solution

Define all arguments as static string variables. This makes arguments more consistent, and cleaner codebase.

mpkorstanje commented on core/src/main/java/io/cucumber/core/options/CommandlineOptions.java:
```diff
@@ -0,0 +1,53 @@
+package io.cucumber.core.options;
+
+public class CommandlineOptions {
```
- This class should be final. 
- Moved to `io.cucumber.core.cli`
- Annotated with `@Api` (same as the Main class)
- Add a private constructor to prevent instantiation

mpkorstanje replied:
And don't forget to add a line to the `CHANGELOG`

ltpquang replied:
@mpkorstanje good point on private constructor 👍 I've updated the PR as your suggestions, please help me review them. Thanks!

mpkorstanje reviewed (changes requested):
This is useful. Since this class is to be used in combination with the main class it would make sense to put it in the same package. It is essentially an interface for public consumption..

mpkorstanje commented on CHANGELOG.md:
```diff
@@ -11,6 +11,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ### Changed
 
+ * [Core] Refactor CommandLine arguments ([2130](https://github.com/cucumber/cucumber-jvm/pull/2130) Quang Lê)
```
The change log is read from the perspective of a user of the library. From that perspective you did not refactor any thing. Rather you added constants for use with the CLI.

ltpquang replied:
yeah I've just updated it for easier to understand 👍

mpkorstanje commented on core/src/main/java/io/cucumber/core/cli/CLIOptions.java:
```diff
@@ -0,0 +1,58 @@
+import org.apiguardian.api.API;
+
+@API(status = API.Status.STABLE)
+public final class CLIOptions {
```
CommandLineOptions was a good name. Why did you change it?

mpkorstanje replied:
And don't forget to add some java doc. Gotta explain to people what this is for. And what to use it in combination with.

ltpquang replied:
I renamed because I didn't use static import at first, making the code a little bit clumsy. Since I'd used static import, I reverted it to the old name. And included some javadoc too. 😄

mpkorstanje commented on core/src/main/java/io/cucumber/core/options/CommandlineOptionsParser.java:
```diff
@@ -65,66 +66,70 @@ private RuntimeOptionsBuilder parse(List<String> args) {
                 exitCode = 0;
                 return parsedOptions;
-            } else if (arg.equals("--version") || arg.equals("-v")) {
+            } else if (arg.equals(CLIOptions.VERSION) || arg.equals(CLIOptions.VERSION_SHORT)) {
```
For readability static imports would be useful

ltpquang replied:
static import is good, I've just updated that

mpkorstanje commented on core/src/main/java/io/cucumber/core/cli/CommandlineOptions.java:
```diff
@@ -0,0 +1,70 @@
+ * <p>
+ * All the options are defined as static string variables to allow other programs to call
+ * {@link io.cucumber.core.cli.Main#main} function in a more consistent way.
+ * E.g. {@code io.cucumber.core.cli.Main.main(NAME, "TestName", THREADS, 2)}
```
Note that Main.main will call ` System.exit(exitStatus)`. This may not be desirable for most users. Would it be worth adding a `Main.run(String... argv)`?

mpkorstanje replied:
Also `"2"`

mpkorstanje reviewed (changes requested):
Almost there!

mpkorstanje commented:
I pushed some changes. Let me know what you you think.

ltpquang commented:
> I pushed some changes. Let me know what you you think.

Thanks @mpkorstanje , everything seems to be great now. But for later PRs, what changes will be categorized as `Added`, and what will be categorized as `Changed`, in the CHANGELOG? Since my PR _changes_ some behavior, by _adding_ some variables, this makes me a little bit confused.

mpkorstanje commented:
We're using Keep a Changelog as a guide: https://keepachangelog.com/en/1.0.0/

> Types of changes
> 
>     Added for new features.
>     Changed for changes in existing functionality.
>     Deprecated for soon-to-be removed features.
>     Removed for now removed features.
>     Fixed for any bug fixes.
>     Security in case of vulnerabilities.

Since this wasn't existing behaviour, changed does not apply.

ltpquang commented:
> We're using Keep a Changelog as a guide: https://keepachangelog.com/en/1.0.0/
> 
> > Types of changes
> > ```
> > Added for new features.
> > Changed for changes in existing functionality.
> > Deprecated for soon-to-be removed features.
> > Removed for now removed features.
> > Fixed for any bug fixes.
> > Security in case of vulnerabilities.
> > ```
> 
> Since this wasn't existing behaviour, changed does not apply.

@mpkorstanje Got it. Everything is clear by now, please continue PR process as I don't have any further comments. Thank you!

aslakhellesoy commented:
Hi @ltpquang,

Thanks for your making your first contribution to Cucumber, and welcome to the Cucumber committers team! You can now push directly to this repo and all other repos under the cucumber organization! 🍾

In return for this generous offer we hope you will:

* ✅ Continue to use branches and pull requests. When someone on the core team approves a pull request (yours or someone else's), you're welcome to merge it yourself.
* 💚 Commit to setting a good example by following and upholding our [code of conduct](https://cucumber.io/conduct) in your interactions with other collaborators and users.
* 💬 Join the [community Slack channel](http://cucumber.io/support#slack) to meet the rest of the team and make yourself at home.
* ℹ️ Don't feel obliged to help, just do what you can if you have the time and the energy.
* 🙋 Ask if you need anything. We're looking for feedback about how to make the project more welcoming, so please tell us!

On behalf of the Cucumber core team,
Aslak Hellesøy
Creator of Cucumber
